### PR TITLE
Joints Awareness: stream out the Cartesian and Image positions of the robot joints so as objects

### PR DESCRIPTION
**Pull request by maxime-petit**
Small module that stream informations about the iCub body:
- read the location of the iCub's joints (left/right arm or both + torso) and stream all of them (not just the end effector).
`/jointsAwareness/<part>/jointsLoc:o`
`Format: x0 y0 z0 x1 y1 z1 ... xn yn zn`

- calculate the projection of these joints into the image from the camera (left or right) and stream them
`/jointsAwareness/<part>/to<camera>Cam/joints2DProj:o`
`Format: x0 y0 x1 y1 ... xn yn`

- read iol2opc streaming object location, compute the projection into the image from the camera (left/right) and stream them. The format is for ABM to be able to store them properly
`/jointsAwareness/objects/to<camera>Cam/objects2DProj:o`
`Format: ((objName0 "x0 y0") (objName1 "x1 y1") ... (objNamen "xn yn"))`

### CODE REVIEW

File: main/src/modules/jointsAwareness/include/jointsAwareness.h [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+#include <map>
+#include <yarp/dev/Drivers.h>
+#include <yarp/dev/CartesianControl.h>
+#include <yarp/dev/GazeControl.h>
+#include <yarp/dev/PolyDriver.h>
+#include "wrdac/clients/icubClient.h"
+
+
+
+using namespace wysiwyd::wrdac;
```
Comment (Tobias-Fischer) [inline main/src/modules/jointsAwareness/include/jointsAwareness.h]: Hi @maxime-petit, please don't use `using namespace` in header files. It's evil!

File: main/src/modules/jointsAwareness/src/jointsAwarness.cpp [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+    unsigned int counter = 0;
+    bool isConnected = false;
+    while(!isConnected && counter < 3){
+        yInfo() << "Trying to connect to " << read_ObjLoc_PortName ;
+        isConnected = Network::connect(objLocStreamPortName, read_ObjLoc_PortName);
+        counter++;
+        Time::delay(1.0);
+    }
+    if(!isConnected){
+               yWarning() << "The port " << objLocStreamPortName << " from iol2opc is needed to stream the 2D proj of objects! jointsAwareness will NOT be able to stream this inforations" ;
```
Comment (Tobias-Fischer) [inline main/src/modules/jointsAwareness/src/jointsAwarness.cpp]: Typo: inforations -> information

File: main/src/modules/jointsAwareness/src/jointsAwarness.cpp [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+               isObjectStreamed = false;
```
Comment (Tobias-Fischer) [inline main/src/modules/jointsAwareness/src/jointsAwarness.cpp]: This variable is unnecessary. You can use `read_ObjLoc_Port.getInputCount() > 0` instead.

File: main/src/modules/jointsAwareness/src/jointsAwarness.cpp [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+            Vector projection2D;
+            string objName = b->get(i).asList()->get(0).toString();
+            string s_objLoc = b->get(i).asList()->get(1).toString();
+            string s_objProj;
+
+            //yDebug() << "Object " << objName << " with loc: " << s_objLoc;
+
+            Bottle bCurrentObjectLoc;
+
+            std::string::size_type sz; // alias of size_t
```
Comment (Tobias-Fischer) [inline main/src/modules/jointsAwareness/src/jointsAwarness.cpp]: Looks like this is only used as optional argument, and then never used. Can be deleted.

File: main/src/modules/jointsAwareness/src/jointsAwarness.cpp [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+            }
+
+            if(camera == "left") {
+                iGaze->get2DPixel(0, vObjLoc, projection2D); //project to left cam
+            } else {
+                iGaze->get2DPixel(1, vObjLoc, projection2D); //project to right cam
+            }
+
+            for(unsigned int p = 0; p < projection2D.size(); p++){   //to be inside ABM, the projection has to be a string with the x/y
+                if(p == 0){                                          //begin with a space otherwrise
```
Comment (Tobias-Fischer) [inline main/src/modules/jointsAwareness/src/jointsAwarness.cpp]: Typo in `otherwrise`

**Comment by Tobias-Fischer** [comment]
Hi @pattacini,
I think that is what @maxime-petit originally did, but as we only record 5 joints or so, I don't think it's worth the hassle. Also, we will never use this module when the full demo is running, so the network should be more or less idle.

Best, Tobi

**Comment by maxime-petit** [comment]
Hi  Ugo, thanks for the quick reply!

For `get2DPixel` I already have some code that reproduce the get2DPixel, in the branch feature/joints3d-to-ks2d, with this commit: f70f36cd560f9972d7edb5862abc456a4635a17d 

I have only done some basic tests by the time, but it seemed to work, so it should not be too long/hard to incorporate it into jointsAwareness. Do you want me to do it before merging the pull request or after?

What do you mean by "intrinsic parameters for the cameras"? I don't control the head at all, I just want the projection, so the left/right are just here to decide in which camera I need to project to, which is more an "outside" decision (i.e. depending on which camera Hyung Jin is using to extract kinematic structure) than an information from gaze controller.

File: main/src/modules/jointsAwareness/include/jointsAwareness.h (hunk 1 above)
Comment (maxime-petit) [inline main/src/modules/jointsAwareness/include/jointsAwareness.h]: Sure :) I was not using them anyway that was a copy/paste from somewhere that I forgot to delete

File: main/src/modules/jointsAwareness/src/jointsAwarness.cpp (hunk 2 above)
Comment (maxime-petit) [inline main/src/modules/jointsAwareness/src/jointsAwarness.cpp]: ack!

File: main/src/modules/jointsAwareness/src/jointsAwarness.cpp (hunk 3 above)
Comment (maxime-petit) [inline main/src/modules/jointsAwareness/src/jointsAwarness.cpp]: Thanks for the tip :)

File: main/src/modules/jointsAwareness/src/jointsAwarness.cpp (hunk 4 above)
Comment (maxime-petit) [inline main/src/modules/jointsAwareness/src/jointsAwarness.cpp]: If it is working without it, sure :D

File: main/src/modules/jointsAwareness/src/jointsAwarness.cpp (hunk 5 above)
Comment (maxime-petit) [inline main/src/modules/jointsAwareness/src/jointsAwarness.cpp]: ack

**Comment by pattacini** [comment]
Hi @maxime-petit,

That has noting to do with the control. You have to populate the **projection matrix** (the Prj matrix [here](https://github.com/robotology/icub-main/blob/master/src/modules/iKinGazeCtrl/src/localizer.cpp#L303)) with the **intrinsic parameters of the cameras**.

However, following @Tobias-Fischer's suggestions you could proceed without taking this issue into account.

**Comment by maxime-petit** [comment]
Ah ok I understand now, that was still with the idea of reproducing `get2DPixel`. I was thinking it was another point because of the "cameraSuffix" :+1: 

Thanks all :)

./Max

**Pull request merged**